### PR TITLE
Implement support for LPUART

### PR DESCRIPTION
**Pull request by FrozenDroid**
Have tested this with BufferedUart, seems to work well. I have not tested uart v1.
Creating this PR now to allow for discussion on whether or not this is the best way to do it.

### CODE REVIEW

File: embassy-stm32/build.rs [hunk 1]
```diff
@@ -255,6 +255,16 @@ fn main() {
         (("usart", "CTS"), quote!(crate::usart::CtsPin)),
         (("usart", "RTS"), quote!(crate::usart::RtsPin)),
         (("usart", "CK"), quote!(crate::usart::CkPin)),
+        (("lpuart", "TX"), quote!(crate::usart::TxPin)),
+        (("lpuart", "RX"), quote!(crate::usart::RxPin)),
+        (("lpuart", "CTS"), quote!(crate::usart::CtsPin)),
+        (("lpuart", "RTS"), quote!(crate::usart::RtsPin)),
+        (("lpuart", "CK"), quote!(crate::usart::CkPin)),
+        (("lpuart", "TX"), quote!(crate::usart::TxPin)),
```
Comment (Dirbaio) [inline embassy-stm32/build.rs:263]: these are duplicated

Comment (FrozenDroid) [reply embassy-stm32/build.rs:263]: Yeah, though the usart also has duplicated ones, so I thought it was for a reason.

File: embassy-stm32/src/usart/mod.rs [hunk 2]
```diff
@@ -455,55 +455,91 @@ unsafe fn clear_interrupt_flags(_r: crate::pac::usart::Usart, _sr: regs::Sr) {
...
 }
 
-pin_trait!(RxPin, Instance);
-pin_trait!(TxPin, Instance);
-pin_trait!(CtsPin, Instance);
-pin_trait!(RtsPin, Instance);
-pin_trait!(CkPin, Instance);
+pub trait BasicUartInstance: sealed::BasicUartInstance {}
+
+pub trait UartInstance: sealed::UartInstance {}
```
[the remark this answers is not in the document]
Comment (FrozenDroid) [reply embassy-stm32/src/usart/mod.rs]: Renamed to BasicInstance and FullInstance

**Comment by FrozenDroid** [comment]
> Have you checked the version of UART and LPUART matches in all chips?
I haven't, how would I go about searching through all the files to check this?

**Comment by Dirbaio** [comment]
manually :smiling_face_with_tear: ... Checking one chip per family should be safe enough.

**Comment by FrozenDroid** [comment]
Ah dear, already found an example of a chip (STM32L4A6) that has a v1 LPUART but a v2 USART...
🙃 

Where do we go from here?

**Comment by darkwater** [comment]
I've been asked to take over here, but I'm not sure how to properly resolve the conflicts. Apparently this is a PR from our master branch, which also seems to have other work on it. Should I open a new PR from a dedicated branch, with just the LPUART commits, rebased on current upstream master?

**Comment by Dirbaio** [comment]
yes! there should be no unrelated commits. If you have to open a new PR then so be it :)

**Pull request merged**